### PR TITLE
Prevent context pollution and scope inline budget in chat

### PR DESCRIPTION
**Pull request by lwc6**
## Summary

This PR replaces full-history replay with a strategy-based context builder.

### Changes
- Added `ContextBuilder` + `ProviderContextPolicy` + `ConversationTurn`
- Applied provider-specific limits for text/image history and inline budget
- Excluded historical failed turns (`Error: ...`) from future context
- Validated inline budget against selected context (not raw full history)
- Updated related tests

### Why
This reduces oversized payloads and prevents failed-turn context pollution.

### Note
The current policy values are still heuristic and somewhat arbitrary.  
If you’re open to it, I’d love to iterate together on a more principled policy

## Summary by CodeRabbit

* **Improvements**
  * Enhanced conversation context assembly with provider-specific policies, retaining recent/current turns and pruning old attachments for leaner requests.
  * Better handling of assistant messages: sanitization of error fragments and exclusion of empty or invalid assistant responses.
  * Improved inline attachment validation to consider prepared conversation turns.

* **Tests**
  * Added and updated tests covering attachment budget validation and assistant-error handling/edge cases.

### CODE REVIEW

File: app/src/main/kotlin/dev/chungjungsoo/gptmobile/util/ApiStateFlowExtensions.kt [hunk 1]
```diff
@@ -164,6 +165,8 @@ internal fun stripAssistantErrorNote(content: String): String {
     }
 }
 
+internal fun isAssistantErrorMessage(content: String): Boolean = content.trimStart().startsWith(ASSISTANT_ERROR_PREFIX)
```
Comment (Taewan-P) [inline app/src/main/kotlin/dev/chungjungsoo/gptmobile/util/ApiStateFlowExtensions.kt]: It seems like that this function should be in a different file. This file is specifically for ApiState Flows extension functions

File: app/src/test/kotlin/dev/chungjungsoo/gptmobile/data/repository/ChatRepositoryImplTest.kt [hunk 2]
```diff
@@ -177,8 +179,58 @@ class ChatRepositoryImplTest {
...
 
+    @Test
+    fun `failed historical turn is excluded from subsequent inline budget checks`() = runBlocking {
+        val openAIAPI = RecordingOpenAIAPI()
+        val repository = createRepository(openAIAPI = openAIAPI)
+        val failedTurnAttachment = ChatAttachment(
+            localFilePath = "Z:/missing/oversized.png",
+            preparedFilePath = "Z:/missing/oversized.png",
+            displayName = "oversized.png",
+            mimeType = "image/png",
```
[the remark this answers is not in the document]
Comment (Taewan-P) [reply app/src/test/kotlin/dev/chungjungsoo/gptmobile/data/repository/ChatRepositoryImplTest.kt]: True. This comment is valid. Please fix this part 🙏

File: app/src/test/kotlin/dev/chungjungsoo/gptmobile/util/ApiStateFlowExtensionsTest.kt
Comment (Taewan-P) [inline app/src/test/kotlin/dev/chungjungsoo/gptmobile/util/ApiStateFlowExtensionsTest.kt:1]: This test location should also change I guess

File: app/src/main/kotlin/dev/chungjungsoo/gptmobile/data/context/ContextBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        userMessages: List<MessageV2>,
+        assistantMessages: List<List<MessageV2>>,
+        platform: PlatformV2,
+        policy: ProviderContextPolicy = ProviderContextPolicy.forClientType(platform.compatibleType)
+    ): List<ConversationTurn> {
+        if (userMessages.isEmpty()) return emptyList()
+
+        val rawTurns = userMessages.mapIndexed { index, userMessage ->
+            val assistantMessage = assistantMessages.getOrNull(index)
+                ?.firstOrNull { it.content.isNotBlank() && it.platformType == platform.uid }
```
[the remark this answers is not in the document]
Comment (Taewan-P) [reply app/src/main/kotlin/dev/chungjungsoo/gptmobile/data/context/ContextBuilder.kt]: I think this makes sense. What do you think?

File: app/src/main/kotlin/dev/chungjungsoo/gptmobile/data/context/ContextBuilder.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
+package dev.chungjungsoo.gptmobile.data.context
+
+import dev.chungjungsoo.gptmobile.data.database.entity.MessageV2
+import dev.chungjungsoo.gptmobile.data.database.entity.PlatformV2
+import dev.chungjungsoo.gptmobile.util.isAssistantErrorMessage
+import javax.inject.Inject
```
[the remark this answers is not in the document]
Comment (Taewan-P) [reply app/src/main/kotlin/dev/chungjungsoo/gptmobile/data/context/ContextBuilder.kt:9]: This will be fixed when you move the function to a different file

**Review by Taewan-P — CHANGES_REQUESTED** [review]
Hello, Thanks for the PR! 

I haven't finished reading it yet, but I did make a few notes about things that could be improved.

**Comment by lwc6** [comment]
Thank you for your suggestion. I have made modifications to the code and submitted it

**Comment by Taewan-P** [comment]
@lwc6 I would like to cross check with you whether the context pollution PR is still valid for the latest update! I added revision feature in #273, and you can switch revisions which can not pollute message context by selecting the successful revision. (Also can you fix the conflicts? Sorry)

**Comment by lwc6** [comment]
@Taewan-P The issue persists. You can try uploading a file larger than 12MB, and then the entire session becomes unusable. If you delete that large uploaded attachment, the app will crash directly. I have modified the relevant code based on your latest version. I hope you can take a look.

File: app/src/test/kotlin/dev/chungjungsoo/gptmobile/data/repository/ChatRepositoryImplTest.kt [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package dev.chungjungsoo.gptmobile.data.repository
 
 import android.content.ContextWrapper
```
Comment (Taewan-P) [inline app/src/test/kotlin/dev/chungjungsoo/gptmobile/data/repository/ChatRepositoryImplTest.kt:3]: Please reformat the import lines

**Review by Taewan-P — COMMENTED** [review]
Other than this, I think it is good to go.

File: app/src/test/kotlin/dev/chungjungsoo/gptmobile/data/repository/ChatRepositoryImplTest.kt (hunk 5 above)
Comment (lwc6) [reply app/src/test/kotlin/dev/chungjungsoo/gptmobile/data/repository/ChatRepositoryImplTest.kt:3]: What's the problem with this?

Comment (Taewan-P) [reply app/src/test/kotlin/dev/chungjungsoo/gptmobile/data/repository/ChatRepositoryImplTest.kt:3]: image

Import order is slightly wrong according to ktlint

**Review by Taewan-P — APPROVED** [review]
LGTM
Thanks for the contribution!
